### ISSUE TITLE
Full base order should be maintained in overlay results

### ISSUE THREAD

**Issue by clehene** [kubernetes-sigs/kustomize#756]
There were several issues and patches related to ordering of build results.

The current behavior doesn't seem to reflect the intended semantics of having a tool that is intended to be used in a similar fashion with git workflows.

- The logical order of elements in a Kubernetes template is not the alphabetical order.
- As a Kustomize user, I need to understand the result of the kustomization by comparing the overlay build results to the base.
- Instead Kutomize outputs results in alphabetical order, without comments

This makes the diff between the build result and the base impracticable. Entire yaml substructures are shifted, comments removed etc.

The only way to have reasonable way to understand the result of the build with diff is to have bases with elements in alphabetical order. That IMHO is inconsistent with the Kustomize core intentions

**Comment by clehene** [comment]
For example trying to see what the effect of a kustomization was:
![image](https://user-images.githubusercontent.com/101597/52147461-a07c8b80-261b-11e9-9c4f-6eeb64bdd265.png)

**Comment by monopole** [comment]
These issue are related, but we might be able to get more help if it were sharded into smaller chunks, and expressed more concretely as "for the following command i want the following output".
E.g. make a small test in https://github.com/kubernetes-sigs/kustomize/tree/master/pkg/target
with an example of comments not being preserved.

**Comment by bai** [comment]
I think with git workflow in mind, ideal behavior would be keeping manifest keys order as is (unchanged).

It kind of makes sense to rearrange Resources (e.g. make sure Namespace created first, ConfigMap created second, Deployment depending on first two created third). Although I could see how this _could_ be little too opinionated for some use cases.

However, within a single Resource sorting keys in alphabetical order does _not_ make sense IMO: this definitely screws up git diffs and goes against the order people are used to after reading countless manifests from Kubernetes documentation and examples.

Being said that, I wanted to suggest the following:

1. By default, do not reorder YAML keys: keep ordering as is. Add an option to `kustomization.yaml` to allow current behavior.
2. (RFC) Add an option do disable resource reordering.

Thoughts?

**Comment by monopole** [comment]
See https://github.com/kubernetes-sigs/kustomize/blob/master/docs/v_2.1.0.md#resource-ordering

**Comment by monopole** [comment]
@bai pretty much did as you suggested. :)

**Comment by bluebrown** [comment]
is there now any solution for reversing the property key order in a given YAML doc? Its hard to follow, for me.
